pkg/providers/gitlab.go

```diff
@@ -60,25 +60,56 @@ func (g *gitLab) Fetch() (*File, error) {
 	}
 
 	candidates := []*assets.Asset{}
-	for _, link := range release.Assets.Links {
-		candidates = append(candidates, &assets.Asset{Name: link.Name, URL: link.URL})
+
+	packages, _, err := g.client.Packages.ListProjectPackages(projectPath, &gitlab.ListProjectPackagesOptions{})
+	if err != nil {
+		return nil, err
 	}
+	packageFound := false
+	for _, v := range packages {
+		if fmt.Sprintf("v%s", v.Version) == release.TagName {
+			packageFiles, _, err := g.client.Packages.ListPackageFiles(projectPath, v.ID, &gitlab.ListPackageFilesOptions{})
+			if err != nil {
+				return nil, err
+			}
+			for _, f := range packageFiles {
+				candidates = append(candidates, &assets.Asset{
+					Name: v.Name,
+					URL: fmt.Sprintf("%sprojects/%s/packages/%s/%s/%s/%s",
+						g.client.BaseURL().String(),
+						url.PathEscape(projectPath),
+						v.PackageType,
+						v.Name,
+						v.Version,
+						f.FileName,
+					),
+				})
+			}
+			packageFound = true
+		}
+	}
+
+	if !packageFound {
```

Comment:
nit: Do we need the packagesFound bool condition? Can't we just check len(candidates) > 0?

Reply:
> Shall we consider both packages and assets as candidates for public repos?

Sounds like a good approach! We could possibly detect the `goreleaser` uploads type of paths combined with the project's visibility setting to determine if the `goreleaser` uploads would be accessible or not. I'll see if it's doable.

Reply:
I've attempted the above in d9937e9.

Reply:
PS: We could additionally choose to only look for generic packages _if no plausibly publicly available assets are found_.

Reply:
Please note that in 028a647, I've altered the asset scoring algorithm so it doesn't give an asset bonus points for mentioning the same desired name components (arch, os, repo name) in _both_ the URL and the name. Previously, such assets might double their score by having those components in both URL and name and would overshadow other candidates. In the case of a GitLab release with both asset URLs and a matching generic package, the generic package would "win" over the asset URL.